### ISSUE TITLE
MA0042 is reported when the async method it proposed to use is the method from within the invocation is done

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#1034]
### Version of the Meziantou.Analyzer NuGet package

3.0.11

### Rule Identifier

MA0042

### Target Framework

.NET 10.0

### C# Language version

C# 14

### Description

MA0042 is currently "wrongfully" reported for a non-async method call when that call is done from within the async method it proposed to use. Applying the "suggestion" would lead to a stackoverflow.

I came across this in legacy code.

### Reproduction Steps

Minimal code:

```c#
public sealed class B
{
    public void Run()
    {
    }

    public async Task RunAsync()
    {
        Run(); // MA0042: Use 'RunAsync' instead of 'Run' | NOK, should not be reported
    }

    public async Task Work()
    {
    }

    public async Task WorkAsync()
    {
        Run(); // MA0042: Use 'RunAsync' instead of 'Run' | OK
    }
}

```

**Comment by meziantou** [comment]
Duplicate of #788. Will improve the documentation

**Comment by drieseng** [comment]
@meziantou, why not fix this issue?

**Comment by meziantou** [comment]
It's complicated to be correct and the number of reports is quite low. So I think manual suppression is ok.
The main question is how you can be sure that replacing `Run` with `RunAsync` method will actually be a recursive call. C# is complicated with attribute, extension methods, C#14 extensions

Example:

```c#
    public void Run()
    {
    }

    public async Task RunAsync()
    {
        await RunAsync(); // Not a recursive call here
    }

    [OverloadPriority(-1)]
    public async Task RunAsync(int i = 0)
    {
    }

```

**Comment by drieseng** [comment]
You're right, thanks for the feedback!
